### PR TITLE
Morello TLS

### PR DESCRIPTION
**Pull request by bsdjhb**

### CODE REVIEW

File: libexec/rtld-elf/aarch64/reloc.c [hunk 1]
```diff
@@ -754,9 +808,38 @@ reloc_non_plt(Obj_Entry *obj, Obj_Entry *obj_rtld, int flags,
...
 			break;
+#ifdef __CHERI_PURE_CAPABILITY__
+		case R_MORELLO_TLSDESC:
+#else
 		case R_AARCH64_TLSDESC:
+#endif
 			reloc_tlsdesc(obj, rela, where, flags, lockstate);
 			break;
+#ifdef __CHERI_PURE_CAPABILITY__
+		case R_MORELLO_TLS_TPREL128:
```
Comment (arichardson) [inline libexec/rtld-elf/aarch64/reloc.c:819]: Add another ifdef in the case to avoid duplicating the comment and the code?

**Review by arichardson — COMMENTED** [review]
Looking forward to testing this 🙂

File: libexec/rtld-elf/aarch64/reloc.c (hunk 1 above)
Comment (bsdjhb) [reply libexec/rtld-elf/aarch64/reloc.c:819]: Yeah, I could.  I think it ends up only being the `where` assignments that differ.

**Comment by arichardson** [comment]
Can we write something to the rootfs so that cheribuild can detect whether it needs to pass `-femulated-tls`? I guess we could parse the libc.so symbol table, but something simpler would be good. Maybe just a version bump in sys/param.h?

File: share/mk/bsd.cpu.mk [hunk 2]
```diff
@@ -304,10 +304,7 @@ MACHINE_CPU += riscv
 
 .if ${MACHINE_CPUARCH} == "aarch64"
 . if ${MACHINE_ARCH:Maarch64*c*}
-# XXX: Stop using emulated TLS once purecap TLSDESC is properly specified (and
-# CheriBSD implements the resolvers) or the Morello toolchain implements a
-# pure-capability traditional TLS like MIPS or RISC-V.
-CFLAGS+=	-march=morello+c64 -mabi=purecap -femulated-tls
+CFLAGS+=	-march=morello+c64 -mabi=purecap
```
Comment (arichardson) [inline share/mk/bsd.cpu.mk:307]: Since this commit breaks the libc ABI, maybe it should also increase the CheriBSD version?

**Comment by bsdjhb** [comment]
This now includes the var-args changes as well and bumps `__CheriBSD_version`

**Review by jrtc27 — COMMENTED** [review]
A couple of style-ish comments; up to you if you want to quickly fix them or get the changes in and have a later cleanup PR come in some time over the coming week or two, the main thing is getting the functional changes in today

File: libexec/rtld-elf/aarch64/reloc.c [hunk 3]
```diff
@@ -775,10 +837,18 @@ reloc_non_plt(Obj_Entry *obj, Obj_Entry *obj_rtld, int flags,
 					return (-1);
 				}
 			}
+#ifdef __CHERI_PURE_CAPABILITY__
+			where[0] = def->st_value + rela->r_addend +
```
[the remark this answers is not in the document]
Comment (bsdjhb) [reply libexec/rtld-elf/aarch64/reloc.c:841]: Yeah, I debated on that one.  The reason I kept this separate is to make it clear the entry has two halves and using *where and where[1] would be a bit odd.  I could perhaps use where[0] always.

File: bin/cheribsdtest/cheribsdtest_tls_threads.c [hunk 4]
```diff
@@ -209,6 +209,28 @@ CHERIBSDTEST(test_tls_threads, "Test TLS across threads")
 #endif
 	CHECK_DISJOINT(tls_ie);
 	CHECK_DISJOINT(tls_le);
+
+#ifdef __aarch64__
```
[the remark this answers is not in the document]
Comment (bsdjhb) [reply bin/cheribsdtest/cheribsdtest_tls_threads.c:213]: Ok

File: libexec/rtld-elf/aarch64/reloc.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (jrtc27) [reply libexec/rtld-elf/aarch64/reloc.c:841]: Yeah, I understand how you end up where you did and am a little unsure of it myself, but the main thing is not wanting to repeat the value+addend+tlsoffset calculation that for me makes it worth merging, for both conciseness and clarity (takes a moment to check the two are in fact the same currently)

File: share/mk/bsd.compat.mk [hunk 5]
```diff
@@ -121,8 +120,7 @@ HAS_COMPAT+=64C
 LIB64C_MACHINE=	arm64
 LIB64C_MACHINE_ARCH=	aarch64c
 LIB64CCPUFLAGS=	-target aarch64-unknown-freebsd13.0
-# XXX: Drop -femulated-tls once bsd.cpu.mk no longer passes it
-LIB64CCPUFLAGS+=	-march=morello+c64 -mabi=purecap -femulated-tls
+LIB64CCPUFLAGS+=	-march=morello+c64 -mabi=purecap
```
Comment (bsdjhb) [inline share/mk/bsd.compat.mk:123]: This file needs the varargs changes?

Comment (jrtc27) [reply share/mk/bsd.compat.mk:123]: bsd.cpu.mk sets it regardless of ABI as it affects both, so it doesn't need repeating here (unless you ever have a vanilla world with lib64c, but we don't support that)

Comment (bsdjhb) [reply share/mk/bsd.compat.mk:123]: Perhaps not since we append these to the CFLAGS from bsd.cpu.mk?

Comment (jrtc27) [reply share/mk/bsd.compat.mk:123]: Having said that we do redundantly set -march for RISC-V, for example, despite the fact that's not changing between ABIs

**Pull request merged**